### DISCUSSION TITLE
[ARM64] Cross compiling detects USE_SSE2 flag on host

### DISCUSSION THREAD

**Discussion by biswadipmaity** [openMVG/openMVG#1876] [Q&A] [answered]
Hi all,
I am trying to cross compile openMVG. Host: x86, Target: aarch64 (NVIDIA Jetson TX2). 
Currently, I am seeing 2 issues:

1.  ```src/nonFree/sift/CMakeLists.txt``` detects `USE_SSE2` flag as true and then make fails because the compilers for ARM don't recognize `emmintrin.h`. 
2. Similarly, `mavx2` flag is unrecognized because `USE_AVX2` flag is set. 

Temporarily I have removed those regions in the CMakeLists.txt to make it work, but I was wondering if there is any other cleaner way to support cross-compilation ? 

PS: I am using this method for cross compilation:
https://www.stereolabs.com/docs/docker/building-arm-container-on-x86/

Thanks for your suggestions!

**Comment by pmoulon** [comment]
Hello @biswadipmaity Thank you for using Github discussion.
Can you list the cmake command line you used for configuring OpenMVG so it would ease reproducibility of your setup?

We are more than willing to help the community to build for ARM platform (I know we made it work in the past but it was a bit of manual configuration) but since more more ARM platform are accessible to anyone nowadays it is good to plan a long term support for it.

**Reply by biswadipmaity** [reply]
Hello @pmoulon Thanks for your help !
The cmake command I am using to configure OpenMVG is: `cmake -DOpenMVG_USE_OCVSIFT=ON ..` 
I have tried adding cmake flags like: `-DUSE_SSE2=OFF`, `-DUSE_SSE2=FALSE`, `-DVL_DISABLE_SSE2=ON`, `-DVL_DISABLE_SSE2` but without success. 

Since I am running cmake inside a qemu based container for arm, I do not need to specify any cross compilation toolchain. 
I think the problem is cmake detects my host as Zen (which is my host), and hence it enables the USE_AVX2, and USE_SSE2 flags. However, these are not supported on the target architecture. I have made the following change as a temporary workaround to make it work:

![OpenMVG](https://user-images.githubusercontent.com/12016122/112744040-4c4a6900-8f51-11eb-9823-1a40e8e564e0.png)

**Comment by pmoulon** [comment, marked as answer]
>  cmake detects my host as Zen (which is my host)

Can you try add `-DTARGET_ARCHITECTURE=generic` in your cmake command line?
It should disable the SSE and AVX option and so let you compile...
Let me know if it works?

**Reply by pmoulon** [reply]
@biswadipmaity Any feedback?

**Reply by biswadipmaity** [reply]
Hi @pmoulon ! Sorry for the late response.. it worked like a charm! Thanks for your help.. :)
